Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp

```diff
@@ -0,0 +1,487 @@
+#include "Acts/EventData/detail/TransformationBoundToFree.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/MagneticField/ConstantBField.hpp"
+#include "Acts/MagneticField/MagneticFieldContext.hpp"
+#include "Acts/Propagator/EigenStepper.hpp"
+#include "Acts/Propagator/Navigator.hpp"
+#include "Acts/Propagator/Propagator.hpp"
+#include "Acts/Propagator/VoidNavigator.hpp"
+#include "Acts/Propagator/detail/JacobianEngine.hpp"
+#include "Acts/Surfaces/DiscSurface.hpp"
+#include "Acts/Surfaces/PlaneSurface.hpp"
+
+#include <cmath>
+#include <cstdlib>
+#include <exception>
+#include <iomanip>
+#include <iostream>
+
+using namespace Acts;
+using namespace Acts::UnitLiterals;
+
+// #define DEBUG_LOG
+
+namespace {
+/** Helper function to compute dt/ds
+ * Helper function to compute the derivative of the time as function of the path
+ * length
+ */
+template <class T_ParticleHypothesis>
+double computeDtDs(const T_ParticleHypothesis &hypothesis, double qop) {
+  return std::hypot(1., hypothesis.mass() / hypothesis.extractMomentum(qop));
+}
+
+/** Compute the path length derivatives for the free/bound to curvilinear
+ * paramter transform.
+ * @param direction the direction of trajectory at the location in question
+ * @param qop q/p of the particle at the location in question in Acts units
+ * @param bfield the magnetic field at the location in question in Acts units
+ * @param particle_hypothesis the particle hypothesis e.g. Acts::ParticleHypothesis::pion()
+ * @return path length derivatives ( dr(...)/ds, dt/ds, dr(...)/ds2, d qop/ds [== 0] )
+ */
+template <class T_ParticleHypothesis>
+FreeToPathMatrix computeFreeToPathDerivatives(
+    const Vector3 &direction, double qop, const Vector3 &bfield,
+    const T_ParticleHypothesis &particle_hypothis) {
+  FreeToPathMatrix path_length_deriv;
+  static_assert(path_length_deriv.cols() ==
+                8);  // ensure that all elements are initialized
+  path_length_deriv.segment<3>(eFreePos0) = direction;
+  path_length_deriv(0, eFreeTime) = computeDtDs(particle_hypothis, qop);
+  path_length_deriv.segment<3>(eFreeDir0) =
+      (qop * direction.cross(bfield)).transpose();
+  path_length_deriv(0, Acts::eFreeQOverP) = 0.;
+  return path_length_deriv;
+}
+template <typename T, std::size_t Rows, std::size_t Cols>
+inline constexpr Eigen::Matrix<T, Rows, Cols> makeMatrix(
+    std::initializer_list<double> elements) {
+  // static_assert( elements.size() == Rows*Cols )
+  if (!(elements.size() == Rows * Cols)) {
+    // throw std::range_error("Initializer list size does not match matrix
+    // dimensions.");
+    std::abort();
+  }
+  Eigen::Matrix<T, Rows, Cols> matrix;
+  auto iter = elements.begin();
+  for (unsigned int row_i = 0; row_i < matrix.rows(); ++row_i) {
+    for (unsigned int col_i = 0; col_i < matrix.cols(); ++col_i) {
+      matrix(row_i, col_i) = *iter;
+      ++iter;
+    }
+  }
+  return matrix;
+}
+
+}  // namespace
+
+struct TestData {
```

Comment:
The test data does not contain the expected output for the covariance matrix in curvilinear parametrisation. It is assumed that the propagator correctly computes it for at least one step size or that the alternative method is correct. However, both methods call the same methods. So, while for one acts version it was correct, the test would not be sensitive to all kinds of changes.

Reply:
There does not seem to be a strong opinion. closing.